client/src/main/scala/client/WebSocketClient.scala

```diff
@@ -1,92 +1,101 @@
 package client
 
-import client.WebSocketClient.WebSocketClientException
-import monifu.reactive.Ack.Continue
-import monifu.reactive.OverflowStrategy.DropOld
-import monifu.reactive._
-import monifu.reactive.channels.PublishChannel
+import client.WebSocketClient.SocketException
+import monix.execution.Ack.{Continue, Stop}
+import monix.execution.cancelables.SingleAssignmentCancelable
+import monix.execution.{Ack, Cancelable}
+import monix.reactive.observers.Subscriber
+import monix.reactive.{Observable, OverflowStrategy}
 import org.scalajs.dom
 import org.scalajs.dom.raw.MessageEvent
 import org.scalajs.dom.{CloseEvent, ErrorEvent, Event, WebSocket}
 
-import scala.concurrent.Promise
+import scala.concurrent.Future
 import scala.concurrent.duration._
+import scala.util.Try
 import scala.util.control.NonFatal
-import scala.util.{Success, Try}
 
-class WebSocketClient(url: String, timeout: Option[FiniteDuration]) extends Observable[String] with (String => Unit) {
-  private val sendOnOpen = Promise[String => Unit]()
-
-  def apply(message: String): Unit = {
-    import scala.concurrent.ExecutionContext.Implicits.global
-    sendOnOpen.future.onComplete {
-      case Success(send) => send(message)
-      case _ => ()
-    }
-  }
-
-  def onSubscribe(subscriber: Subscriber[String]): Unit = {
-    import subscriber.scheduler
-
-    def inboundWrapper(webSocket: WebSocket) = try {
-      val inbound = PublishChannel[String](DropOld(2))
-      webSocket.onopen = (event: Event) => {
-        // todo close properly
-        val outbound = PublishChannel[String](DropOld(2)).sample(Observable.intervalAtFixedRate(100.millis, 1.second))
-        outbound.subscribe { e =>
-          webSocket.send(e)
-          Continue
+class WebSocketClient private(url: String, messages: Either[String, () => Observable[String]]) extends Observable[String] {
+
+  def unsafeSubscribeFn(subscriber: Subscriber[String]): Cancelable = {
```

Comment:
I want to refactor with such contract for `SuiteExecution`.
```
trait SuiteExecution {
  /**
    * Executes tests in this <code>Suite</code>.
    *
    * @param channel is used for imperative style feeding of output events (test result)
    */
  def executionTestSuite(suite: Suite, channel: String => Unit, closeChannel: () => Unit = { () => () }): Unit = {
    val s = new OutputStream() {
      override def write(b: Array[Byte], off: Int, len: Int) =
        if (len != 0) channel(new String(b, off, len))

      override def write(b: Int): Unit = channel(b.toChar.toString)
    }
    try {
      //execute() - ScalaTest entry point
      Console.withOut(s)(suite.execute())
      closeChannel()
    } finally s.close()
  }
}
```
But I can't. `7420e40` introduced additional `StringBuilder`:
1) eating server's memory in case of huge test output
2) changing `side effect only contract` to `side effect + result` contract which not easy to refactor/adopt/change
```
object StringBuilderRunner {
  def apply(block: => (String => Unit) => String,
            testResult: Try[String] => Option[TestResult] = { _ => None })
           (implicit ec: ExecutionContext): String = {
    val sb = new StringBuilder
    block(sb.append(_))
    testResult(Success(sb.toString())).map(_.testStatus.toString).foreach(sb.append)
    sb.toString()
  }
}
Please look at reference solution for such case:
```
object StringBuilderRunner {
  def apply(block: => (String => Unit) => Unit)(implicit ec: ExecutionContext): String = {
    val sb = new StringBuilder
    block(s => sb.append(s))
    sb.toString()
  }
}
```
But we definitely can't waste server's memory as in `7420e40`. Initial goal was to push `test output` to client as soon as possible, can we revert this?